craft_application/services/lifecycle.py

```diff
@@ -268,6 +282,22 @@ def __repr__(self) -> str:
             f"{work_dir=}, {cache_dir=}, {build_for=}, **{self._manager_kwargs!r})"
         )
 
+    def _update_project_metadata(self) -> None:
+        """Replace project fields with values adopted during the lifecycle."""
+        self._update_project_variables()
+
+    def _update_project_variables(self) -> None:
+        """Replace project fields with values set using craftctl."""
+        update_vars: dict[str, str] = {}
+        for var in self._app.project_variables:
+            value = self._lcm.project_info.get_project_var(var)
+            if not value:
+                raise errors.PartsLifecycleError(f"Project field '{var}' was not set.")
+            update_vars[var] = value
+
+        emit.debug(f"Update project variables: {update_vars}")
+        self._project = self._project.copy(update=update_vars)
```

Comment:
This will also need to update the project that's accessed by the PackageService. It should work fine to update the project attached to the ServiceFactory for that, but we should definitely do an integration test to check that.

Reply:
I think it's risky to set a new Project object - even if we set if through the service factory, it the Package service has already been initialized then it's too late. It also means that there will be other Project objects around (like in the Application itself) that are stale/invalid. I guess letting the project be mutable might be the lesser of two evils here.

I definitely agree that this needs to be integration-tested, though. Maybe the fake PackageService used in the application integration tests can use the project's version when setting the name of the package (currently it's just "package.tar.zst"